Blog on fine-tuning a model with KAITO (#4484)
Azure/AKS: blog/_posts/2024-08-19-KAITO-finetuning.md

---
title: "Fine tune language models with KAITO on AKS"
description: "Make your AI model even smarter, in just a few steps"
date: 2024-08-19
author: Sachi Desai

> categories: AI, Machine Learning, open-source, operator

Reviewer: You need to pick a single category here. We need to add something for AI. I'll send details in Teams

> categories: AI, Machine Learning, open-source, operator

Author: ```suggestion
categories: AI
```

---

> **Introduction**

Author: ```suggestion
## Introduction
```

All you have to do is specify your model preset and tuning method, data set, and container registry details - KAITO will automate the rest and provide a lightweight output to pull into new inferencing jobs!

**Fine-tune a Phi-3 model with KAITO**

Let’s look at what fine-tuning looks like with KAITO. We’ll use Phi-3-mini for our sample deployment. The `Phi-3-mini-128K` model is a robust and high-performing option that’s also a convenient size for fine-tuning, so we’ll use that as an example. With KAITO v0.3.0 or higher, there is an additional tuning property added to the `Workspace` specification. Within the tuning property, you can specify additional information for the tuning job such as a preset configuration for the base model, fine-tuning method using LoRA or QLoRA, input datasets, and output location for the adapter.

> Here’s an example of what the fine-tuning workspace looks like when training on a dataset found on HuggingFace:

Author: ```suggestion
![Screenshot of untuned model response](/blog/assets/images/kaito_untuned_phi3_response.png)
*Screenshot of un-tuned model response*
```

> 

Author: ```suggestion

Uh oh! That is **NOT** the answer we were expecting – we’d expect the model to know more about Kubernetes and cloud technologies. This foundation model was trained on a mix of synthetic data and public websites, so it doesn’t recognize AKS as Azure Kubernetes Service, and got creative in answering the question to the best of its ability. This is a well-known problem in the AI community and various fine-tuning techniques have been invented to enhance model accuracy using domain-specific knowledge. In our specific use case, we need the Phi-3 model to have more knowledge of Kubernetes and cloud platforms.

## KAITO can help with model fine-tuning

KAITO supports Low-Rank Adaption (LoRA), the most common parameter-efficient fine-tuning technique, which significantly lowers the GPU requirement for tuning a language model. With LoRA, fine-tuning an LLM in a single GPU is possible! 

Let’s start the journey of fine-tuning the Phi-3 model using KAITO.

First, we need an input dataset, so we’ll use this [public dataset](https://huggingface.co/datasets/ishaansehgal99/kubernetes-reformatted-remove-outliers) that is optimized for knowledge about Kubernetes and cloud platforms. To use this dataset in KAITO, we can specify it in the new tuning property in the workspace API. 

Here’s an example of what the fine-tuning workspace looks like when using a public dataset from HuggingFace:
```

```bash
apiVersion: kaito.sh/v1alpha1
kind: Workspace
metadata:
  name: workspace-tuning-phi-3
resource:
  instanceType: Standard_NC24ads_A100_v4
  labelSelector:
    matchLabels:
      app: tuning-phi-3
tuning:
  preset:
    name: phi-3-mini-128k-instruct
  method: qlora
  input:
    urls:
      - https://huggingface.co/datasets/myuser/mydataset/resolve/main/data/train-00000-of-00001.parquet?download=true
  output:
    image: myregistry.azurecr.io/adapters/myadapter:0.0.1
    imagePushSecret: myregistrysecret
```

> For the most part, the `Workspace` resource is similar as before. Here, we see that KAITO recommends a minimum GPU VM SKU of “Standard_NC24ads_A100_v4” for the tuning job.  You can update the `instanceType` to a larger GPU VM SKU, but either way, you should make sure there’s enough GPU quota in your Azure subscription prior to deploying applying the workspace to your AKS cluster. 

Author: ```suggestion
The tuning property specifies the minimum information required for starting a tuning job, including:

**Preset**: indicates the target model for tuning (KAITO uses preset configurations to generate all Kubernetes sources needed to run the Job). 
**Method**: users can choose LoRA or QLoRA (quantized LoRA, for lower GPU memory usage) as the tuning method. 
**Input**: for flexibility, the tuning input can be a URL (for a public dataset), or images (for a private dataset).
**Output**: where the adapter is stored, as a container image or any other storage type supported by Kubernetes.

> The choice of GPU SKU is critical since model fine-tuning normally requires more GPU memory compared to model inference. To avoid GPU Out-Of-Memory, NVIDIA A100 or higher tier GPUs are recommended. 

You can check out the [KAITO tuning API](https://github.com/Azure/kaito/tree/main/docs/tuning) for more details and troubleshooting tips!
```

> > Note: Be sure to check out this [quickstart guide](https://learn.microsoft.com/azure/quotas/quickstart-increase-quota-portal) on how to request quota increase in the Azure Portal, if you have never done that before.

Author: ```suggestion
```

> Fine-tuning jobs will run using default tuning configurations defined by KAITO; however, you can optionally create a ConfigMap in your cluster and reference it as your config in the tuning spec. You can configure default parameters to change your fine-tuning result as outlined in this doc: https://github.com/Azure/kaito/tree/main/docs/tuning#categorized-key-parameters

Author: ```suggestion
After deploying the tuning workspace, KAITO will create a Job workload in the same namespace as the workspace and run to completion. A common perception is that model training job would take a long time – well, yes it can be multiple hours even for fine-tuning. You can track the tuning progress in the Job pod log, reported by the number of steps completed out of the total. 

> If you are curious to know how the total steps are calculated, here is the formula: 
> `total steps = number of epochs * (number of samples in dataset / batch size)`, where the number of epochs and batch size are configurable parameters. 

KAITO allows users to apply a custom ConfigMap to overwrite most of the tuning parameters used by the tuning job. We should be careful about changing those parameters though. For example, increasing the batch size, i.e., leveraging higher data parallelism, will reduce the tuning time but use more GPU memory, and require better GPUs in general. 

```

> > Note: If you choose to pull and push from private registries, you must create a Kubernetes “docker-registry” secret and reference it as your `imagePullSecret`/`imagePushSecret`.

Author: ```suggestion
When the fine-tuning job is complete, the result, which is often referred to as an **adapter**, will be packaged as a container image and stored in the specified output location. The adapter image is lightweight, portable, conveniently version controlled, and can be pulled into a new inferencing service!

After submitting your tuning workspace manifest, go get yourself a drink or a snack. 😉

![Three hours later](/blog/assets/images/three_hours_later_image.png)
```

> After deploying the KAITO workspace, KAITO will create a `Job` workload in the same namespace as the workspace and run to completion. This can take several hours depending on the size of your input model and dataset.  Once the fine-tuning job is complete, an adapter result will be stored in the output location specified in your workspace tuning output settings. This adapter layer is stored as a container image, holding a subset of updated weights of your model based on what it learned from your input dataset. As a result, the image is lightweight, portable, conveniently version controlled, and can be pulled into new inferencing jobs!

Author: ```suggestion
## Now that tuning is done, how can we quickly see the result?

```

> To use your new fine-tuned adapter, the inference workspace can now accept one or more adapters whose strength values can be adjusted (default of 1.0) to customize model behavior. 

Author: ```suggestion
To use our new fine-tuned adapter, the inference workspace will now accept one or more adapters to customize model behavior.
```

Here is an example of what a fine-tuned inference workspace looks like:

```bash
apiVersion: kaito.sh/v1alpha1
kind: Workspace
metadata:
  name: workspace-phi-3-mini-adapter
resource:
  instanceType: Standard_NC6s_v3
  labelSelector:
    matchLabels:
      apps: phi-3-adapter
inference:
  preset:
    name: phi-3-mini-128k-instruct
  adapters:
    - source:
        name: myadapter
        image: myregistry.azurecr.io/adapters/myadapter:0.0.1
        imagePullSecrets:
          - myregistrysecret
      strength: "1.0"
```

> As new inference requests come in, they'll flow through the model and adapters so the response will be altered based on new training data and adjusted based on the strength of each adapter. 

Author: ```suggestion
As new inference requests come in, they’ll flow through the model that merges the adapters and the response will be affected by the newly trained data. 
Great! After updating our inference endpoint to leverage the new adapter, let’s ask the same question and check the response from the fine-tuned phi-3-mini-128k-instruct model:
```

> Here is an example of what an un-tuned `phi-3-mini-128k-instruct` model response looks like, versus a fine-tuned `phi-3-mini-128k-instruct` model:

Author: ```suggestion
```

> ![Screenshot of untuned model response](/blog/assets/images/kaito_untuned_phi3_response.png)

Author: ```suggestion
```

*Screenshot of un-tuned model response*

![Screenshot of tuned model response](/blog/assets/images/kaito_tuned_phi3_response.png)
*Screenshot of tuned model response*

> You can see in the two screenshots above, that the base `phi-3-mini-128k-instruct` model with no fine-tuning wasn’t quite sure of what the user means by “AKS”. However, with a little bit of fine-tuning with a [dataset](https://huggingface.co/datasets/ishaansehgal99/kubernetes-reformatted-remove-outliers) optimized for Kubernetes and cloud platforms, it is quickly able to discern that the user is looking for an answer related to Kubernetes and provides a better response. There is still a bit of work to do when it comes to prompt engineering and/or fine-tuning tweaks but it’s clearly on the right path!

Author: ```suggestion
That definitely looks more accurate – the fine-tuned model quickly discerned that we’re looking for an answer related to Kubernetes and provided a much better response. However, there’s always room to improve the model’s accuracy for domain specific questions. We can improve upon or find new input datasets to create new tuning adapters. This iterative tuning process can be done by repeating the steps described above.

## Other options to make models smarter

Retrieval-Augmented Generation (RAG) is another common technique used to improve the inference accuracy of foundation models. Compared to LoRA fine-tuning, RAG eliminates the need of training Jobs to generate adapters. It has a more complicated workflow with additional components like a vector database, indexing and query servers, etc. The good news is that RAG support with a simple user experience is in KAITO’s roadmap. 🚀
```

> **Summary**

Author: ```suggestion
## Summary
```

> With KAITO’s new fine-tuning capability, you can now give open-source models additional training for specific tasks, which will generate more context-aware results to your business needs. KAITO supports small language model (SLM) families like Phi-3, which you can choose to reduce costs and fine-tune quickly in comparison to larger, general-purpose models. The best part is that as new data flows in, you can iterate with multiple adapters to improve model performance and achieve this sort of ML workflow all within your AKS cluster and with just a few lines of YAML!

Author: ```suggestion
Users expect LLMs to stay on task and provide accurate answers. While foundation models on their own have a wealth of knowledge, they often need to be refined for specific tasks. This can be achieved by leveraging techniques like fine-tuning or RAG, but the process of implementing these techniques is not always straightforward. As you saw, with KAITO, it becomes **really easy**. All you need to do is specify an input dataset, a GPU SKU, and off it goes! KAITO will take care of the rest. Its adapter design makes it very portable to attach to multiple workspaces or even attach multiple adapters to a single workspace. In the end, we can see that the inference and tuning workspaces along with adapters can become fundamental building blocks in your overall ML workflow.
```

> We’re just getting started and would love your feedback. To learn more about KAITO fine-tuning, AI model deployment on AKS or KAITO’s roadmap in general, check out the following links.

Author: ```suggestion
We’re just getting started and would love your feedback. To learn more about KAITO fine-tuning, AI model deployment on AKS or KAITO’s roadmap in general, check out the following links:
```

To use your new fine-tuned adapter, the inference workspace can now accept one or more adapters whose strength values can be adjusted (default of 1.0) to customize model behavior. 

Here is an example of what a fine-tuned inference workspace looks like:

```bash
apiVersion: kaito.sh/v1alpha1
kind: Workspace
metadata:
  name: workspace-phi-3-mini-adapter
resource:
  instanceType: Standard_NC6s_v3
  labelSelector:
    matchLabels:
      apps: phi-3-adapter
inference:
  preset:
    name: phi-3-mini-128k-instruct
  adapters:
    - source:
        name: myadapter
        image: myregistry.azurecr.io/adapters/myadapter:0.0.1
        imagePullSecrets:
          - myregistrysecret
      strength: "1.0"
```

As new inference requests come in, they’ll flow through the model that merges the adapters and the response will be affected by the newly trained data. 
Great! After updating our inference endpoint to leverage the new adapter, let’s ask the same question and check the response from the fine-tuned phi-3-mini-128k-instruct model:

> *Screenshot of un-tuned model response*

Author: ```suggestion
```

---
title: "Fine tune language models with KAITO on AKS"
description: "Make your AI model even smarter, in just a few steps"

> date: 2024-08-19

Author: ```suggestion
date: 2024-08-23
```

> Note: Be sure to check out this [quickstart guide](https://learn.microsoft.com/azure/quotas/quickstart-increase-quota-portal) on how to request quota increase in the Azure Portal, if you have never done that before.

Fine-tuning jobs will run using default tuning configurations defined by KAITO; however, you can optionally create a ConfigMap in your cluster and reference it as your config in the tuning spec. You can configure default parameters to change your fine-tuning result as outlined in this doc: https://github.com/Azure/kaito/tree/main/docs/tuning#categorized-key-parameters

When the fine-tuning job is complete, the result, which is often referred to as an **adapter**, will be packaged as a container image and stored in the specified output location. The adapter image is lightweight, portable, conveniently version controlled, and can be pulled into a new inferencing service!

After submitting your tuning workspace manifest, go get yourself a drink or a snack. 😉

![Three hours later](/blog/assets/images/three_hours_later_image.png)

## Now that tuning is done, how can we quickly see the result?

> 

Author: ```suggestion
```

---
title: "Fine tune language models with KAITO on AKS"
description: "Make your AI model even smarter, in just a few steps"
date: 2024-08-23
author: Sachi Desai
categories: AI
---

## Introduction

You may have heard of the [Kubernetes AI Toolchain Operator](https://github.com/Azure/kaito/tree/main) (KAITO) announced at Ignite 2023 and KubeCon Europe this year. The open source project has gained popularity in recent months by introducing a streamlined approach to AI model deployment and flexible infrastructure provisioning on Kubernetes.

With the [v0.3.0 release](https://github.com/Azure/kaito/releases/tag/v0.3.0), KAITO has expanded the supported model library to include the Phi-3 model, but the biggest (and most exciting) addition is the ability to fine-tune open-source models. Why should you be excited about fine-tuning? Well, it’s because fine-tuning is one way of giving your foundation model additional training using a specific dataset to enhance accuracy, which ultimately improves the interaction with end-users.

> 

Author: ```suggestion
```

We’re just getting started and would love your feedback. To learn more about KAITO fine-tuning, AI model deployment on AKS or KAITO’s roadmap in general, check out the following links:

**Resources**

- [KAITO fine-tuning guide](https://github.com/Azure/kaito/tree/main/docs/tuning)
- [Concepts - Fine-tuning language models for AI and machine learning workflows](https://learn.microsoft.com/azure/aks/concepts-fine-tune-language-models)
- [Concepts - Small and large language models](https://learn.microsoft.com/azure/aks/concepts-ai-ml-language-models)

> - [KAITO Roadmap](https://github.com/orgs/Azure/projects/669)

Author: ```suggestion
- [KAITO Roadmap](https://github.com/orgs/Azure/projects/669)

## Acknowledgements
Huge thanks to [Paul Yu](https://www.linkedin.com/in/yupaul) for co-authoring this blog and providing valuable feedback on KAITO since the early stages of the project!
```

![Screenshot of untuned model response](/blog/assets/images/kaito_untuned_phi3_response.png)
*Screenshot of un-tuned model response*

Uh oh! That is **NOT** the answer we were expecting – we’d expect the model to know more about Kubernetes and cloud technologies. This foundation model was trained on a mix of synthetic data and public websites, so it doesn’t recognize AKS as Azure Kubernetes Service, and got creative in answering the question to the best of its ability. This is a well-known problem in the AI community and various fine-tuning techniques have been invented to enhance model accuracy using domain-specific knowledge. In our specific use case, we need the Phi-3 model to have more knowledge of Kubernetes and cloud platforms.

## KAITO can help with model fine-tuning

KAITO supports Low-Rank Adaption (LoRA), the most common parameter-efficient fine-tuning technique, which significantly lowers the GPU requirement for tuning a language model. With LoRA, fine-tuning an LLM in a single GPU is possible! 

Let’s start the journey of fine-tuning the Phi-3 model using KAITO.

> First, we need an input dataset, so we’ll use this [public dataset](https://huggingface.co/datasets/ishaansehgal99/kubernetes-reformatted-remove-outliers) that is optimized for knowledge about Kubernetes and cloud platforms. To use this dataset in KAITO, we can specify it in the new tuning property in the workspace API. 

Author: ```suggestion
First, we need an input dataset, so we’ll use a [public dataset](https://huggingface.co/datasets/myuser/mydataset/resolve/main/data/train-00000-of-00001.parquet?download=true) that is optimized for knowledge about Kubernetes and cloud platforms. (Note: this is a pseudo-link to a personal data location, where the original data set was copied and reformatted in a personal HuggingFace account.) 

To use the data set in KAITO, we can specify it in the new `tuning` property in the workspace API. 
```

Here’s an example of what the fine-tuning workspace looks like when using a public dataset from HuggingFace:
```bash
apiVersion: kaito.sh/v1alpha1
kind: Workspace
metadata:
  name: workspace-tuning-phi-3
resource:
  instanceType: Standard_NC24ads_A100_v4
  labelSelector:
    matchLabels:
      app: tuning-phi-3
tuning:
  preset:
    name: phi-3-mini-128k-instruct
  method: qlora
  input:
    urls:
      - https://huggingface.co/datasets/myuser/mydataset/resolve/main/data/train-00000-of-00001.parquet?download=true
  output:
    image: myregistry.azurecr.io/adapters/myadapter:0.0.1
    imagePushSecret: myregistrysecret
```

The tuning property specifies the minimum information required for starting a tuning job, including:

> **Preset**: indicates the target model for tuning (KAITO uses preset configurations to generate all Kubernetes sources needed to run the Job). 

Author: ```suggestion
**Preset**: indicates the target model for tuning (KAITO uses preset configurations to generate all Kubernetes sources needed to run the Job). 

```

> **Method**: users can choose LoRA or QLoRA (quantized LoRA, for lower GPU memory usage) as the tuning method. 

Author: ```suggestion
**Method**: users can choose LoRA or QLoRA (quantized LoRA, for lower GPU memory usage) as the tuning method. 

```

> **Input**: for flexibility, the tuning input can be a URL (for a public dataset), or images (for a private dataset).

Author: ```suggestion
**Input**: for flexibility, the tuning input can be a URL (for a public dataset), or images (for a private dataset).

```

When the fine-tuning job is complete, the result, which is often referred to as an **adapter**, will be packaged as a container image and stored in the specified output location. The adapter image is lightweight, portable, conveniently version controlled, and can be pulled into a new inferencing service!

After submitting your tuning workspace manifest, go get yourself a drink or a snack. 😉

![Three hours later](/blog/assets/images/three_hours_later_image.png)

## Now that tuning is done, how can we quickly see the result?

To use our new fine-tuned adapter, the inference workspace will now accept one or more adapters to customize model behavior.

Here is an example of what a fine-tuned inference workspace looks like:

```bash
apiVersion: kaito.sh/v1alpha1
kind: Workspace
metadata:
  name: workspace-phi-3-mini-adapter
resource:
  instanceType: Standard_NC6s_v3
  labelSelector:
    matchLabels:
      apps: phi-3-adapter
inference:
  preset:
    name: phi-3-mini-128k-instruct
  adapters:
    - source:
        name: myadapter
        image: myregistry.azurecr.io/adapters/myadapter:0.0.1
        imagePullSecrets:
          - myregistrysecret

>       strength: "1.0"

Author: ```suggestion
```

  method: qlora
  input:
    urls:
      - https://huggingface.co/datasets/myuser/mydataset/resolve/main/data/train-00000-of-00001.parquet?download=true
  output:
    image: myregistry.azurecr.io/adapters/myadapter:0.0.1
    imagePushSecret: myregistrysecret
```

The tuning property specifies the minimum information required for starting a tuning job, including:

**Preset**: indicates the target model for tuning (KAITO uses preset configurations to generate all Kubernetes sources needed to run the Job). 
**Method**: users can choose LoRA or QLoRA (quantized LoRA, for lower GPU memory usage) as the tuning method. 
**Input**: for flexibility, the tuning input can be a URL (for a public dataset), or images (for a private dataset).
**Output**: where the adapter is stored, as a container image or any other storage type supported by Kubernetes.

> The choice of GPU SKU is critical since model fine-tuning normally requires more GPU memory compared to model inference. To avoid GPU Out-Of-Memory, NVIDIA A100 or higher tier GPUs are recommended. 

You can check out the [KAITO tuning API](https://github.com/Azure/kaito/tree/main/docs/tuning) for more details and troubleshooting tips!

> 

Author: ```suggestion
```

After deploying the tuning workspace, KAITO will create a Job workload in the same namespace as the workspace and run to completion. A common perception is that model training job would take a long time – well, yes it can be multiple hours even for fine-tuning. You can track the tuning progress in the Job pod log, reported by the number of steps completed out of the total. 

> If you are curious to know how the total steps are calculated, here is the formula: 
> `total steps = number of epochs * (number of samples in dataset / batch size)`, where the number of epochs and batch size are configurable parameters. 

KAITO allows users to apply a custom ConfigMap to overwrite most of the tuning parameters used by the tuning job. We should be careful about changing those parameters though. For example, increasing the batch size, i.e., leveraging higher data parallelism, will reduce the tuning time but use more GPU memory, and require better GPUs in general. 

> 

Author: ```suggestion
```

![Three hours later](/blog/assets/images/three_hours_later_image.png)

## Now that tuning is done, how can we quickly see the result?

To use our new fine-tuned adapter, the inference workspace will now accept one or more adapters to customize model behavior.

Here is an example of what a fine-tuned inference workspace looks like:

```bash
apiVersion: kaito.sh/v1alpha1
kind: Workspace
metadata:
  name: workspace-phi-3-mini-adapter
resource:
  instanceType: Standard_NC6s_v3
  labelSelector:
    matchLabels:
      apps: phi-3-adapter
inference:
  preset:
    name: phi-3-mini-128k-instruct
  adapters:
    - source:
        name: myadapter
        image: myregistry.azurecr.io/adapters/myadapter:0.0.1
        imagePullSecrets:
          - myregistrysecret
      strength: "1.0"
```

As new inference requests come in, they’ll flow through the model that merges the adapters and the response will be affected by the newly trained data. 
Great! After updating our inference endpoint to leverage the new adapter, let’s ask the same question and check the response from the fine-tuned phi-3-mini-128k-instruct model:

> 

Author: ```suggestion
```

## Fine-tuning a foundation model is sometimes necessary

Let’s experiment with [Phi-3-mini-128K](https://huggingface.co/microsoft/Phi-3-medium-128k-instruct), a robust and high-performing model that’s a bit smaller in size than your average LLM. We’ll use the following [chatbot UI tool](https://streamlit.io/), connected to a Phi-3 inference service deployed by KAITO, and ask a basic question about AKS:

![Screenshot of untuned model response](/blog/assets/images/kaito_untuned_phi3_response.png)
*Screenshot of un-tuned model response*

Uh oh! That is **NOT** the answer we were expecting – we’d expect the model to know more about Kubernetes and cloud technologies. This foundation model was trained on a mix of synthetic data and public websites, so it doesn’t recognize AKS as Azure Kubernetes Service, and got creative in answering the question to the best of its ability. This is a well-known problem in the AI community and various fine-tuning techniques have been invented to enhance model accuracy using domain-specific knowledge. In our specific use case, we need the Phi-3 model to have more knowledge of Kubernetes and cloud platforms.

## KAITO can help with model fine-tuning

> KAITO supports Low-Rank Adaption (LoRA), the most common parameter-efficient fine-tuning technique, which significantly lowers the GPU requirement for tuning a language model. With LoRA, fine-tuning an LLM in a single GPU is possible! 

Author: ```suggestion
KAITO supports Low-Rank Adaption (LoRA) and Quantized LoRA (QLoRA), the [most common](https://arxiv.org/pdf/2402.16141) parameter-efficient fine-tuning techniques, which significantly lower the GPU requirement for tuning a language model. With LoRA, fine-tuning an LLM in a single GPU is possible! 
```

Let’s start the journey of fine-tuning the Phi-3 model using KAITO.

> First, we need an input dataset, so we’ll use a [public dataset](https://huggingface.co/datasets/myuser/mydataset/resolve/main/data/train-00000-of-00001.parquet?download=true) that is optimized for knowledge about Kubernetes and cloud platforms. (Note: this is a pseudo-link to a personal data location, where the original data set was copied and reformatted in a personal HuggingFace account.) 

Author: ```suggestion
First, we need an input dataset, so we’ll use a [public dataset](https://huggingface.co/datasets/sidddddddddddd/kubernetes) that is optimized for knowledge about Kubernetes and cloud platforms. (Note: this is the original data set which we copied and reformatted using [these instructions](https://github.com/Azure/kaito/tree/main/docs/tuning#input-dataset-format) from the HuggingFace trainer library.) 
```

## Fine-tuning a foundation model is sometimes necessary

Let’s experiment with [Phi-3-mini-128K](https://huggingface.co/microsoft/Phi-3-medium-128k-instruct), a robust and high-performing model that’s a bit smaller in size than your average LLM. We’ll use the following [chatbot UI tool](https://streamlit.io/), connected to a Phi-3 inference service deployed by KAITO, and ask a basic question about AKS:

![Screenshot of untuned model response](/blog/assets/images/kaito_untuned_phi3_response.png)
*Screenshot of un-tuned model response*

Uh oh! That is **NOT** the answer we were expecting – we’d expect the model to know more about Kubernetes and cloud technologies. This foundation model was trained on a mix of synthetic data and public websites, so it doesn’t recognize AKS as Azure Kubernetes Service, and got creative in answering the question to the best of its ability. This is a well-known problem in the AI community and various fine-tuning techniques have been invented to enhance model accuracy using domain-specific knowledge. In our specific use case, we need the Phi-3 model to have more knowledge of Kubernetes and cloud platforms.

## KAITO can help with model fine-tuning

> KAITO supports Low-Rank Adaption (LoRA), the most common parameter-efficient fine-tuning technique, which significantly lowers the GPU requirement for tuning a language model. With LoRA, fine-tuning an LLM in a single GPU is possible! 

Reviewer 2: Naive reader questions:
* any link/literature to support this is the most common technique? 
* will we consider supporting others?

Let’s start the journey of fine-tuning the Phi-3 model using KAITO.

> First, we need an input dataset, so we’ll use a [public dataset](https://huggingface.co/datasets/myuser/mydataset/resolve/main/data/train-00000-of-00001.parquet?download=true) that is optimized for knowledge about Kubernetes and cloud platforms. (Note: this is a pseudo-link to a personal data location, where the original data set was copied and reformatted in a personal HuggingFace account.) 

Reviewer 2: hmm, you are hyperlinking a link that needs to be changed, when people click that it will always fail. Shouldn't we instead copy it in full with the places to change clearly listed? or alternatively hyperlink to a place that will help you get to the dataset

  labelSelector:
    matchLabels:
      apps: phi-3-adapter
inference:
  preset:
    name: phi-3-mini-128k-instruct
  adapters:
    - source:
        name: myadapter
        image: myregistry.azurecr.io/adapters/myadapter:0.0.1
        imagePullSecrets:
          - myregistrysecret
      strength: "1.0"
```

As new inference requests come in, they’ll flow through the model that merges the adapters and the response will be affected by the newly trained data. 
Great! After updating our inference endpoint to leverage the new adapter, let’s ask the same question and check the response from the fine-tuned phi-3-mini-128k-instruct model:

![Screenshot of tuned model response](/blog/assets/images/kaito_tuned_phi3_response.png)
*Screenshot of tuned model response*

That definitely looks more accurate – the fine-tuned model quickly discerned that we’re looking for an answer related to Kubernetes and provided a much better response. However, there’s always room to improve the model’s accuracy for domain specific questions. We can improve upon or find new input datasets to create new tuning adapters. This iterative tuning process can be done by repeating the steps described above.

> ## Other options to make models smarter

Reviewer 2: To my question above, this might warrant an in-page link from above. And no strict need to put this in the end if it makes more sense when you introduce the method

We’re just getting started and would love your feedback. To learn more about KAITO fine-tuning, AI model deployment on AKS or KAITO’s roadmap in general, check out the following links:

**Resources**

- [KAITO fine-tuning guide](https://github.com/Azure/kaito/tree/main/docs/tuning)
- [Concepts - Fine-tuning language models for AI and machine learning workflows](https://learn.microsoft.com/azure/aks/concepts-fine-tune-language-models)
- [Concepts - Small and large language models](https://learn.microsoft.com/azure/aks/concepts-ai-ml-language-models)
- [KAITO Roadmap](https://github.com/orgs/Azure/projects/669)

## Acknowledgements

> Huge thanks to [Paul Yu](https://www.linkedin.com/in/yupaul) for co-authoring this blog and providing valuable feedback on KAITO since the early stages of the project!

Reviewer 2: Nothing against this, could we just have co-author alternatively? More curious on the option

Author: @palma21 I checked with Paul on this, and the blog as a GitHub page doesn't support having co-authors
